### PR TITLE
Add deregistration email jobs

### PR DESCRIPTION
**Pull request by endofunky**

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -90,6 +91,8 @@
     t.text "deregistration_message"
     t.date "deregistered_at"
     t.index ["exemption_id"], name: "index_registration_exemptions_on_exemption_id"
+    t.index ["registration_id"], name: "i_rex_batch", where: "((state)::text = 'active'::text)"
+    t.index ["registration_id"], name: "index_active_registration_ids_on_registration_exemptions", where: "((state)::text = 'active'::text)"
```
Comment (PaulDoyle-EA) [inline db/schema.rb:94]: Duplicate index?

File: spec/jobs/send_deregistration_email_batch_job_spec.rb [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    ActiveJob::Base.queue_adapter = old_queue_adapter
+  end
+
+  before do
+    allow(WasteExemptionsEngine::FeatureToggle)
+      .to receive(:active?).and_return(feature_toggle_state)
+  end
+
+  describe "#perform" do
+    context "with a valid registrations" do
```
Comment (PaulDoyle-EA) [inline spec/jobs/send_deregistration_email_batch_job_spec.rb]: Minor: typo

File: spec/jobs/send_deregistration_email_batch_job_spec.rb [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+      before { registrations }
+
+      it "enqueues the correct number of email jobs" do
+        job.perform
+
+        expect(SendDeregistrationEmailJob).to have_been_enqueued.exactly(test_batch_size)
+      end
+
+      it "drops the crated temporary table" do
```
Comment (PaulDoyle-EA) [inline spec/jobs/send_deregistration_email_batch_job_spec.rb]: Minor: typo

File: db/schema.rb (hunk 1 above)
Comment (endofunky) [reply db/schema.rb:94]: That was a left over from experimenting in `psql`. Remove now.

File: app/jobs/send_deregistration_email_batch_job.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+              LEFT JOIN registration_exemptions rex
+                  ON r.id = rex.registration_id
+                  AND rex.state = 'active'
+                  AND rex.expires_on - interval '#{renewal_window} days' > '#{today}'
+        WHERE rex.id IS NOT NULL
+          AND r.submitted_at < '#{min_submitted_at}'
+          AND r.deregistration_email_sent_at IS NULL;
+    SQL
+
+    conn.execute("SELECT id FROM #{TEMP_TABLE_NAME} TABLESAMPLE system_rows(#{batch_size})").map do |row|
```
[the remark this answers is not in the document]
Comment (endofunky) [reply app/jobs/send_deregistration_email_batch_job.rb:43]: False positive. Also added to `config/brakeman.ignore`.

File: app/jobs/send_deregistration_email_batch_job.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    conn.execute("DROP TABLE IF EXISTS #{TEMP_TABLE_NAME}")
+
+    conn.execute <<-SQL
+      CREATE TEMP TABLE #{TEMP_TABLE_NAME} AS
+        SELECT DISTINCT r.id AS id
+          FROM registrations r
+              LEFT JOIN registration_exemptions rex
+                  ON r.id = rex.registration_id
+                  AND rex.state = 'active'
+                  AND rex.expires_on - interval '#{renewal_window} days' > '#{today}'
```
[the remark this answers is not in the document]
Comment (endofunky) [reply app/jobs/send_deregistration_email_batch_job.rb:37]: False positive. Also added to `config/brakeman.ignore`.

**Review by PaulDoyle-EA — DISMISSED** [review]

**Review by PaulDoyle-EA — DISMISSED** [review]

**Review by PaulDoyle-EA — APPROVED** [review]

**Pull request merged**